### ISSUE TITLE
[Feature Request]: proposal: decimal(fixed point numeric) data type

### ISSUE THREAD

**Issue by broccoliSpicy** [matrixorigin/matrixone#1867]
### MatrixOne's decimal data type proposal

our expected behavior:

```sql
mysql> create table deci2(a decimal(5, 3));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into deci2 values(12.2);
Query OK, 1 row affected (0.00 sec)

mysql> select * from deci2;
+--------+
| a      |
+--------+
| 12.200 |
+--------+
1 row in set (0.00 sec)

mysql>
mysql> insert into deci2 values(122.2);
ERROR 1264 (22003): Out of range value for column 'a' at row 1
mysql> insert into deci2 values(12.345);
Query OK, 1 row affected (0.00 sec)

mysql> insert into deci2 values(12.3456);
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> select * from deci2;
+--------+
| a      |
+--------+
| 12.200 |
| 12.345 |
| 12.346 |
+--------+
3 rows in set (0.00 sec)
```

MatrixOne's decimal type: decimal(precision, scale)
supported precision range: 0 ~ 38
supported scale range: 0 ~ 38
scale <= precision
supported operation: 
+, -, *, / between decimals, between decimal and integers.
comparison operation <, >, <=, >=, = between decimals, between decimals and integers. 

for the actual behavior in cases decimal operation overflows, further investigation needed.

```go
// Vector's type field 
type Type struct {
	Oid       T
	Size      int32 // e.g. int8.Size = 1, int16.Size = 2, char.Size = 24(SliceHeader size)

	Width     int32
	
	Scale   int32  // the scale value for decimal type
}
    
type decimal0 int64 // precision 1~18

type decimal1 struct { // precision 19~38
	 DecimalValue   [2]int64
}
```

decimal(8, 5) |              representation
--------------| ----------------------:
0.1           |     10000
1.23          |     123000
123.4         |     12340000
123.4567      |     12345670

decimal(25, 10)|             representation:   using [2]decimalValue to emulate int128 
--------------| ----------------------:
0.1           |1,000,000,000       
1.23          |1,230,000,000       
123.4         |1,234,000,000       
123.4567      |1,234,567,000       
123456789.123456789       | 1234567891234567890  
-123456789.123456789       | -1234567891234567890   

for our internal representation, there are only two kinds of precisions: 18(decimal0), 38(decimal1)

operations on this representation:
  SQL standard has specific rules regarding operations of decimal type.
  ```
a) Let S1 and S2 be the scale of the first and second operands respectively.
b) The precision of the result of addition and subtraction is implementation-defined, and the scale is the
maximum of S1 and S2.
c) The precision of the result of multiplication is implementation-defined, and the scale is S1 + S2.
d) The precision and scale of the result of division are implementation-defined.
```

our operation implementation:
addition/substraction:
steps |decimal(8, 5) | operation | decimal(5, 2)
----|------: | :--------: | -----:
-|123.45678|     +    |123.45
representation|12345678 |     +    |12345
align to the maximum scale|12345678|+ |12345000
perform integer addition|
resultDecimal(18, 5) |   | |    24690678
addition result precision: max(operand0.precision, operand1.precision)
addition result scale: max(operand0.scale, operand1.scale)

addition/substraction on decimal1([2]int64): emulate int128 arithmetic (caution!)
decimal + integer: function integerToDecimal
decimal + float:  not defined

multiplication:
steps |decimal(8, 5) | operation | decimal(5, 2)
----|------: | :--------: | -----:
...|123.45678|     *    |123.45
representation|12345678 |     *    |12345
perform integer multiplication|
resultDecimal(38, 7) |   | |   152407394910 

multiplication result precision: 38(decimal1)
multiplication result scale: operand0.scale + operand1.scale

multiplication on decimal1([2]int64): emulate int128 arithmetic (caution!)
decimal * integer: function integerToDecimal
decimal * float:  not defined

division:
steps |decimal(8, 5) | operation | decimal(5, 2)
----|------: | :--------: | -----:
...|123.45678|     /    |123.45
representation|12345678 |     /    |12345
scale dividend using divisor's scale | 1234567800| / | 12345
perform integer division|
resultDecimal(38, 5) |   | |   100005

division result precision: 38(decimal1)
division result scale: dividend's scale

division on decimal1([2]int64): emulate int128 arithmetic (caution!)
decimal / integer: function integerToDecimal
decimal / float:  not defined

string to decimal
decimal to string

implementation:
1. parser
2. computation layer
3. emulate int128 arithmetic using 2 int64
4. storage
        
flaws: 
1. another diverge from mysql, mysql suppports decimal precision up to 81, whereas us only 38
2. the implementation challenge for operations on this decimal representation(emulating int128 arithmetic on [2]int64)
3. compatible issue, I will only emulate int128 arithmetic on x86_64, arm_64.

**Comment by broccoliSpicy** [comment]
another thought from @aunjgr: how about we using [2]int64 to store two separate decimal then concatenate it using string concatenation...
ponding on it...

**Comment by broccoliSpicy** [comment]
> another thought from @aunjgr: how about we using [2]int64 to store two separate decimal then concatenate it using string concatenation...
> ponding on it...

Well, I think the result would be wrong if we treat [2]int64 as two separate decimal value.

the whole emulation method based on the cpu characteristic that during int64 binary multiplication, the binary overflowed is caught, whereas the decimal overflow won't.

**Comment by aunjgr** [comment]
Splitting the whole number by decimal digits instead of binary, ADD/SUB are still trivial, but MUL/DIV have to be intricately implemented.

The possible advantage I believed is that the itoa/atoi conversion would be much more efficient. However this is not true. The compiler doing conversion between int64 and string is still not efficient, i.e. [gcc](https://github.com/gcc-mirror/gcc/blob/16e2427f50c208dfe07d07f18009969502c25dc8/libquadmath/printf/_itoa.h) and [go](https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/internal/itoa/itoa.go).

Probably binary splitting should be better than decimal splitting.

**Comment by broccoliSpicy** [comment]
As for itoa, ClickHouse uses some fancy optimization even compiler won't bother to do, [look up table for itoa, article](https://www.facebook.com/notes/facebook-engineering/three-optimization-tips-for-c/10151361643253920/), [clickhouse source code](https://github.com/ClickHouse/ClickHouse/blob/master/base/base/itoa.h)
for the time being, I think I will stick with the division by 10 method for itoa until further optimization needed

**Comment by JinHai-CN** [comment]
The function will be released in 0.4.0.
